deku-derive/src/lib.rs

```diff
@@ -24,77 +26,309 @@ impl Default for EndianNess {
+    /// if failed.
+    fn from_receiver(receiver: DekuReceiver) -> Result<Self, TokenStream> {
+        // Validate
+        DekuData::validate(&receiver)
+            .map_err(|(span, msg)| syn::Error::new(span, msg).to_compile_error())?;
+
+        let data = match receiver.data {
+            ast::Data::Struct(fields) => ast::Data::Struct(ast::Fields {
+                style: fields.style,
+                fields: fields
+                    .fields
+                    .into_iter()
+                    .map(FieldData::from_receiver)
+                    .collect::<Result<Vec<_>, _>>()?,
+            }),
+            ast::Data::Enum(variants) => ast::Data::Enum(
+                variants
+                    .into_iter()
+                    .map(VariantData::from_receiver)
+                    .collect::<Result<Vec<_>, _>>()?,
+            ),
+        };
 
-        // Validate that `id_type` is set with a size
-        if (self.id_bits.is_some() || self.id_bytes.is_some()) && self.id_type.is_none() {
-            panic!("`id_type` must be specified with `id_bits` or `id_bytes`");
-        }
+        let ctx = receiver
+            .ctx
+            .map(|s| s.parse_with(syn::punctuated::Punctuated::parse_terminated))
+            .transpose()
+            .map_err(|e| e.to_compile_error())?;
+
+        let id_bits = receiver.id_bytes.map(|b| b * 8).or(receiver.id_bits);
+
+        Ok(Self {
+            vis: receiver.vis,
+            ident: receiver.ident,
+            generics: receiver.generics,
+            data,
+            endian: receiver.endian,
+            ctx,
+            id_type: receiver.id_type,
+            id_bits,
+        })
+    }
 
-        // Validate either `id_bits` or `id_bytes` is specified
-        if self.id_bits.is_some() && self.id_bytes.is_some() {
-            panic!("conflicting: both \"id_bits\" and \"id_bytes\" specified on field");
+    fn validate(receiver: &DekuReceiver) -> Result<(), (proc_macro2::Span, &str)> {
+        match receiver.data {
+            ast::Data::Struct(_) => {
+                // Validate id_* attributes are being used on an enum
+                if receiver.id_type.is_some() {
+                    Err((receiver.id_type.span(), "`id_type` only supported on enum"))
+                } else if receiver.id_bytes.is_some() {
+                    Err((
+                        receiver.id_bytes.span(),
+                        "`id_bytes` only supported on enum",
+                    ))
+                } else if receiver.id_bits.is_some() {
+                    Err((receiver.id_bits.span(), "`id_bits` only supported on enum"))
+                } else {
+                    Ok(())
+                }
+            }
+            ast::Data::Enum(_) => {
+                // Validate either `id_bits` or `id_bytes` is specified
+                if (receiver.id_bits.is_some() || receiver.id_bytes.is_some())
+                    && receiver.id_type.is_none()
```

Comment:
I think we could validate `id_type` is set here for enums, and unwrap the option in deku-read/write. That way we're validating in one location

Reply:
Yes, but I think put all validating code in one place is more readable. And we can call `unwrap` in `emit_enum`.

Reply:
Ah I understand what you mean. I will add this validation.

Reply:
Yes exactly, the validation can be done here and then unwrap later 👍